pkg/ingester/tailer.go

```diff
@@ -0,0 +1,231 @@
+		sendChan:       make(chan *logproto.Stream, 2),
+		conn:           conn,
+		droppedStreams: []*logproto.DroppedStream{},
+		id:             generateUniqueID(orgID, query, regex),
+	}, nil
+}
+
+func (t *tailer) loop() {
+	var stream *logproto.Stream
+	var err error
+
+	for {
+		if t.closed {
+			return
+		}
+
+		stream = <-t.sendChan
+		if stream == nil {
+			t.close()
+			return
+		}
+
+		// while sending new stream pop lined up dropped streams metadata for sending to querier
+		tailResponse := logproto.TailResponse{Stream: stream, DroppedStreams: t.popDroppedStreams()}
+		err = t.conn.Send(&tailResponse)
+		if err != nil {
+			level.Error(util.Logger).Log("Error writing to tail client", fmt.Sprintf("%v", err))
+			t.close()
+			return
+		}
+	}
+}
+
+func (t *tailer) send(stream logproto.Stream) {
+	if t.closed {
+		return
+	}
+
+	// if we are already dropping streams due to blocked connection, drop new streams directly to save some effort
+	if blockedSince := t.blockedSince(); blockedSince != nil {
+		if blockedSince.Before(time.Now().Add(-time.Second * 15)) {
+			t.close()
+			close(t.sendChan)
+			return
+		}
+		t.dropStream(stream)
+		return
+	}
+
+	if t.regexp != nil {
+		// filter stream by regex from query
+		t.filterEntriesInStream(&stream)
+		if len(stream.Entries) == 0 {
+			return
+		}
+	}
+
+	select {
+	case t.sendChan <- &stream:
+	default:
+		t.dropStream(stream)
+	}
+}
+
+func (t *tailer) filterEntriesInStream(stream *logproto.Stream) {
+	filteredEntries := new([]logproto.Entry)
+	for _, entry := range stream.Entries {
+		if t.regexp.MatchString(entry.Line) {
+			*filteredEntries = append(*filteredEntries, entry)
+		}
+	}
+
+	stream.Entries = *filteredEntries
+}
+
+func (t *tailer) isWatchingLabels(metric model.Metric) bool {
+	var labelValue model.LabelValue
+	var ok bool
+
+	for _, matcher := range t.matchers {
+		labelValue, ok = metric[model.LabelName(matcher.Name)]
+		if !ok {
+			return false
+		}
+
+		if !matcher.Matches(string(labelValue)) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (t *tailer) isClosed() bool {
+	return t.closed
+}
+
+func (t *tailer) close() {
+	t.closed = true
+}
+
+func (t *tailer) blockedSince() *time.Time {
+	t.blockedMtx.RLock()
+	defer t.blockedMtx.RUnlock()
+
+	return t.blockedAt
+}
+
+func (t *tailer) dropStream(stream logproto.Stream) {
+	t.blockedMtx.Lock()
+	defer t.blockedMtx.Unlock()
+
+	blockedAt := new(time.Time)
+	*blockedAt = time.Now()
+	t.blockedAt = blockedAt
+	t.droppedStreams = append(t.droppedStreams, breakDroppedStream(stream)...)
+}
+
+func (t *tailer) popDroppedStreams() []*logproto.DroppedStream {
+	t.blockedMtx.Lock()
+	defer t.blockedMtx.Unlock()
+
+	if t.blockedAt == nil {
+		return nil
+	}
+
+	droppedStreams := t.droppedStreams
+	t.droppedStreams = []*logproto.DroppedStream{}
+	t.blockedAt = nil
+
+	return droppedStreams
+}
+
+func (t *tailer) getID() uint32 {
+	return t.id
+}
+
+func breakDroppedStream(stream logproto.Stream) []*logproto.DroppedStream {
```

Comment:
This is an extra optimisation, can be removed. It's super unlikely that a user will have >10K streams in a single ingester.